### DISCUSSION TITLE
User-defined / custom subdomain routing

### DISCUSSION THREAD

**Discussion by amfeng** [vercel/next.js#17345] [Help] [answered]
Hi! 

I have a Next.js app that primarily has statically generated pages (using both fallback & incremental generation) and I'd like to have user-defined subdomains in addition to allowing users to point a custom domain to their page. While still maintaining the benefits of having these pages be statically generated and served off a CDN, which means using server-side rendering[0] or client-side content swapping[1] wouldn't be ideal. 

e.g. Currently I have:
```
foo.com/s/:slug => renders statically generated page at /pages/s/[slug]
```

And I'd also like for users to be able to access that page from:
```
custom.foo.com
sub.mycustomdomain.com (via CNAME record to custom.foo or similar)
```

So far the only solution I've found is to use a custom server[2], but I'd love to be able to use Vercel. Other hosts I've chatted with will serve _static sites_ off a CDN, but not _statically generated assets_ on a server (i.e. Next.js run as a Node.js server somewhere else) whereas Vercel's built for that[3].

Ideally I'd be able to hook in somewhere in the routing, check the subdomain and/or custom domain, query a database, then choose one of my statically generated pages to render. Curious what the Next.js or Vercel folks think of that use case?

0: https://github.com/vercel/next.js/discussions/14636
1: https://vercel.com/blog/wildcard-domain
2: https://github.com/vercel/next.js/tree/canary/examples/custom-server-express
3: https://nextjs.org/docs/deployment#optimized-for-nextjs

**Comment by leerob** [comment]
@lachlanjc might have some suggestions 🤞

**Reply by lachlanjc** [reply]
Thanks for the tag! Yep, for sure. We ran into this same problem with the [Hack Club Scrapbook](https://scrapbook.hackclub.com/), using incremental SSG for the profile pages but wanting custom domains. The solution I came up with has a number of compromises—it’s limited by Vercel’s limit of 50 production domains (for pro plans). (Our solution was to make custom domains a limited offer for the first 50 folks who requested, lol.)

Basically, we kept the Next site as-is, then made a separate project for serving the domains that is simply a Vercel serverless function that looks up the host, gets the HTML of the page from the Next site, then serves it from the function. This means a slight performance penalty per-request from custom domains, but I couldn’t find any other relatively straightforward way to do it. That function is then deployed as a separate project, using Vercel’s production domains feature to add folks’ domains via the Vercel Platform API.

- I wrote about the architecture more here, in the “Custom domains” section: https://notebook.lachlanjc.com/2020-07-30_how_scrapbook_works/
- All of Scrapbook is open source: http://github.com/hackclub/summer-scrapbook
- The domains project is open source here: https://github.com/hackclub/summer-domains

**Reply by amfeng** [reply]
@lachlanjc - Thanks so much for the reply!

Interesting, does this mean that the custom domains do not take advantage of the static pages being served by a CDN (since you're effectively serving the HTML from your "server" post-lookup)? I'm also curious how you're grabbing the HTML for each page -- are you going and reading off the build files? Does Next.js count these as "requests" such that it triggers the incremental regeneration?

I also didn't know about the 50-domain limit, that's unfortunate!

**Reply by lachlanjc** [reply]
That is correct, though you can add caching headers on the response. The fetching is just getting the HTML from the live site using fetch, nothing fancy with build files. That will trigger regeneration though, yes. (We use SWR so even if regeneration isn’t up to date on load, the client will update regardless.)

**Reply by trankhacvy** [reply]
@lachlanjc I really like your approach but I sill have some concerns:
- How can you handle these links `<link rel="stylesheet" href="/_next/static/css/xxxxxx.css" >`. It's relative path so it will append to your domain server and return 404.
- If I have a domain, such as `github.com`, how to handle sub-routing like `github.com/name-of-article`

**Comment by jhvissotto** [comment]
I have the same problem!

But I was caching the responses (backend and frotend) with the same library: https://www.npmjs.com/package/apicache

Unfortunately, the problem occors only in frontend (Next.js) that mix my subdomains "www" and "app".

I tryied create two different Redis clients, but the Next mix everything. Not work.

```javascript
var isApp = null
var isNext = null

server.use((req, res, next) => {
    isApp = req.subdomains.includes('app')
    isNext = req.originalUrl.startsWith('/_next')
    return next()
})

if (isNext) {
    server.use(routes)
} else {
    if (isApp) {
        server.use(middleware.cache.app(), routes)
    } else {
        server.use(middleware.cache.web(), routes)
    }
}
```
![image](https://user-images.githubusercontent.com/19745638/94353649-db7dcf00-0049-11eb-95ca-961800dad186.png)

This should be urgently improved.

**Comment by kennethacohen** [comment]
Apologies for bringing this back to life...but just curious if other folks have found a better path forward for these user-defined subdomains.

**Reply by imjakechapman** [reply]
I'm gonna sub to this in case someone responds lol. I'm about to start a project that will need this. Hopefully someone responds with some new insight.

**Reply by surjithctly** [reply]
https://multi-domain-first.vercel.app/

**Reply by steven-tey** [reply]
@kennethacohen @imjakechapman I just posted an example for this in my comment below, hope it helps!

**Comment by trankhacvy** [comment]
Is there anything new on this problem ? I really need this feature on my app.

**Reply by bilalesi** [reply]
Hi, me too i want to domainize my app, is there any solution adapted by Vercel or anyone who experience this feature??

**Reply by nyn0x** [reply]
@trankhacvy @bilalesi try this but  is not perfect
https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225

**Reply by robertpiosik** [reply]
> @trankhacvy @bilalesi try this but is not perfect
> [#12848 (comment)](https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225)

This actually answers the question, why it's not perfect?

**Reply by nyn0x** [reply]
> > @trankhacvy @bilalesi try this but is not perfect
> > [#12848 (comment)](https://github.com/vercel/next.js/discussions/12848#discussioncomment-680225)
> 
> This actually answers the question, why it's not perfect?

because catch all doesn't work ([[...slug]]) with `{...destination: '/:host/:slug*',...}`

**Reply by trankhacvy** [reply]
@nyn0x As my understanding, it only solve the custom subdomain routing problem, right ? how about user-defined domain problem ? any ideas ?

**Reply by robertpiosik** [reply]
@trankhacvy it works for the whole host, not just a subdomain.

**Reply by melissafzhang** [reply]
Is this solution still limited by Vercel's subdomain limit of 50? Or can you use wildcard domains?

**Reply by steven-tey** [reply]
@trankhacvy @bilalesi @melissafzhang I just posted an example for this in my comment below, hope it helps!

**Reply by melissafzhang** [reply]
@steven-tey thanks! Super helpful. In these examples, do we still need to add the custom domain to Vercel (i.e. does the 50 domain limit still apply)?

**Reply by steven-tey** [reply]
@melissafzhang yes you will be adding them to your Vercel project – or if you're letting your users add them, you can add them programmatically using the Domains API (we'll be posting more info on that in the [early access program](https://eap.vercel.sh/)). As for the domain limit, we'll have some updates about that soon, but if you need it increased urgently, feel free to [reach out to us](https://vercel.com/contact/sales) and we'll make sure this is not a blocker for you! :)

**Reply by lachlanjc** [reply]
@sampoder just published an article on this topic as well: https://www.smashingmagazine.com/2021/11/nextjs-wildcard-subdomains/

**Reply by melissafzhang** [reply]
@steven-tey I'm trying to understand how the custom domain example works. What exactly needs to be added using the Domains API?

**Reply by steven-tey** [reply]
@melissafzhang Great question! I just created a simple boilerplate example here that is basically a functioning replica of the domains dashboard on every Vercel project: https://domains-api.vercel.app/

The code is part of the EAP though so feel free to join it [here](https://eap.vercel.sh/) and I'll fast-track you so you can access the repo!

**Reply by steven-tey** [reply]
@lachlanjc love that article! Quick feedback – you mentioned in one of the caveats that:

> No More Static-Site Generation (and ISR): Unfortunately, there aren’t any special systems to provide custom statically generated pages for different wildcard subdomains like you may with dynamic routing for example (where you have [slug].js files).

That isn't true – with the Edge functions [here](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites), you can actually still use SSG/ISR because you're rewriting to a dynamic route and performing static generation on those routes. Do you mind changing that + making some edits to the GH repo example that you created at the end? Thank you!

**Reply by lachlanjc** [reply]
I'm aware that it's incorrect, but I didn't write the article, sorry! Looping in @sampoder :)

**Reply by melissafzhang** [reply]
> @melissafzhang Great question! I just created a simple boilerplate example here that is basically a functioning replica of the domains dashboard on every Vercel project: https://domains-api.vercel.app/
> 
> The code is part of the EAP though so feel free to join it [here](https://eap.vercel.sh/) and I'll fast-track you so you can access the repo!

Thanks! Is there a way to map a domain to a subdomain? e.g. custom-domain.com maps to custom.domain.com. I see you're doing some mapping in `getStaticProps` but wondering if there's a way to do this with DNS

**Reply by steven-tey** [reply]
@melissafzhang yup! it's basically the same thing as mapping subdomains – you'll basically map custom domains to their respective records in the db. First, you distinguish it [here in the middleware file](https://github.com/vercel/examples/blob/main/edge-functions/hostname-rewrites/pages/_middleware.ts#L11-L14), and then inside [getStaticProps](https://github.com/vercel/examples/blob/main/edge-functions/hostname-rewrites/pages/_sites/%5Bsite%5D/index.tsx#L110-L123) you can get the respective data from the database.

**Reply by melissafzhang** [reply]
> @melissafzhang yup! it's basically the same thing as mapping subdomains – you'll basically map custom domains to their respective records in the db. First, you distinguish it [here in the middleware file](https://github.com/vercel/examples/blob/main/edge-functions/hostname-rewrites/pages/_middleware.ts#L11-L14), and then inside [getStaticProps](https://github.com/vercel/examples/blob/main/edge-functions/hostname-rewrites/pages/_sites/%5Bsite%5D/index.tsx#L110-L123) you can get the respective data from the database.

Thanks. Super helpful. Do you have thoughts on how to do client side routing? We have both custom domains and default domains e.g. 
custom.example.com/store and example.com/custom/store. I'm having trouble figuring out the best way to account for both.

**Reply by rosswaycaster** [reply]
> The code is part of the EAP though so feel free to join it [here](https://eap.vercel.sh/) and I'll fast-track you so you can access the repo!

@steven-tey I'm interested in accessing this repo as well! 😃 Just signed up for the Platforms on Vercel EAP!

**Reply by steven-tey** [reply]
@rosswaycaster just added you to the repo! :)

**Reply by melissafzhang** [reply]
@steven-tey would love to be added as well! Thanks!

**Reply by steven-tey** [reply]
@melissafzhang feel free to sign up for the EAP [here](https://eap.vercel.sh) and I'll add you to the repo! :)

**Reply by melissafzhang** [reply]
@steven-tey done!

**Reply by steven-tey** [reply]
@melissafzhang Added you! :)

**Reply by viperfx** [reply]
Just requested too @steven-tey

**Reply by Borales** [reply]
> That isn't true – with the Edge functions here, you can actually still use SSG/ISR because you're rewriting to a dynamic route and performing static generation on those routes.

@steven-tey do you have a working sample for this? without `getServerSideProps` my pages won't work, hence I'm unable to use SSG (otherwise the client-side navigation breaks). 

btw, I also requested EAP.

**Comment by pepas24** [comment]
Hi I'm new in Next.js and I'm trying something similar, I found this repo by Vercel featuring the wildcard domains feature, [https://github.com/vercel/static-fun](https://github.com/vercel/static-fun)
I want to know if this really generate static files on demand or if this is server side rendered, also if it is posible to mask a sub.domain.com with customerdomain.com with some kind of dns redirection. I'm new in this server side things so maybe nothing has sense 😓.

**Reply by robertpiosik** [reply]
You can check if a webpage comes from static file or is server-side rendered by examining its html source code. Right click anywhere on the page and select "View page source" or something similar in your browser and check if page contents are there.

**Reply by byyoungjin** [reply]
it's client-side rendering with useEffect in index.tsx, you can see it renders different page from reading window.location.ref

**Comment by steven-tey** [comment]
Hey @amfeng! We recently rolled out a [simple example](https://github.com/vercel/examples/tree/main/edge-functions/hostname-rewrites) that allows you to create unique content pages with a multi-tenant infrastructure using [edge function](https://vercel.com/features/edge-functions) rewrites and [ISR](https://vercel.com/docs/concepts/next.js/incremental-static-regeneration). Each of your users gets assigned a unique subdomain when they create their account, with the (usually paid) option to add a custom domain. 

We're also rolling out more complex boilerplates (e-commerce storefront builders, website builders) + supplementary code snippets (static tweets, domains API, OG image generator) in an [early access program](https://eap.vercel.sh/) – feel free to join if you're interested (click "sign up" on the Platforms on Vercel program).

**Reply by ntkoopman** [reply]
Is there any way to get custom error pages working?

**Reply by frolic** [reply]
I'd like to sign up for the early access program but wary of the "Act on your behalf" GitHub scope. Why is this needed?

**Reply by steven-tey** [reply]
@ntkoopman we're working on that! :)

**Comment by steven-tey** [comment, marked as answer]
Quick update on this – we recently launched the Platforms Starter Kit that's a comprehensive template for building multi-tenant applications with built-in custom domains support.: https://demo.vercel.pub/platforms-starter-kit

Feel free to try it out and let us know what you think! :)

**Reply by steven-tey** [reply]
CC: @amfeng @pepas24 @melissafzhang @viperfx @Borales @trankhacvy @holic @ntkoopman @kennethacohen @jhenriquevissotto

**Reply by dhruvbhatia7** [reply]
For custom domains, if I already have subdomains working with NextJS wildcard domains, do I need the edge function rewrites? Can't I just ask the user to CNAME their custom domain to my subdomain? For example,

1. My platform is at foo.app
2. I give user a custom subdomain -> company1.foo.app
3. Can the user now add a CNAME record from their custom domain (company1.com to company1.foo.app) WITHOUT edge function rewrite?

**Reply by minardimedia** [reply]
This looks great just was start working on implementing this functionality for my app. 🙌
